File: baekjoon/Java/P1890_점프.java
/*
algorithm: dp
주의점: -2^63 ~ 2^63+1 범위 내에서는 long을 사용함에 주의!
*/
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class P_1890 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[][] arr = new int[N][N];

        long[][] dp = new long[N][N]; //dp 배열 초기화
        dp[0][0] = 1;

        for (int i = 0; i < N; i++) { //배열 입력받기
            String[] line = br.readLine().split(" ");
            for (int j = 0; j < N; j++) {
                arr[i][j] = Integer.parseInt(line[j]);
            }
        }

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if(dp[i][j] > 0 && arr[i][j] > 0) {
                    int jump = arr[i][j];
                    if (i + jump < N) {
                        dp[i + jump][j] += dp[i][j];
                    }
                    if (j + jump < N) {
                        dp[i][j + jump] += dp[i][j];
                    }
                }
            }
        }
        System.out.println(dp[N - 1][N - 1]);
    }
}

File: baekjoon/Python/P1890_점프.py
# dp
import sys
input = sys.stdin.readline

N = int(input())
arr = [list(map(int, input().split())) for _ in range(N)] #입력받기

# dp 생성 & 초기값 생성
dp = [[0]*N for _ in range(N)]
dp[0][0] = 1

for i in range(N):
    for j in range(N):
        if dp[i][j]>0 and arr[i][j]>0: # 경로O, 점프 숫자O
            jump = arr[i][j]
            if i + jump < N: #아래쪽 범위이내
                dp[i + jump][j] += dp[i][j] # 경우수 더해주기
            if j + jump < N:
                dp[i][j + jump] += dp[i][j]
print(dp[N - 1][N - 1])
